pallets/market/src/lib.rs

```diff
@@ -0,0 +1,358 @@
+// This file is part of the SORA network and Polkaswap app.
+
+// Copyright (c) 2020, 2021, Polka Biome Ltd. All rights reserved.
+// SPDX-License-Identifier: BSD-4-Clause
+
+// Redistribution and use in source and binary forms, with or without modification,
+// are permitted provided that the following conditions are met:
+
+// Redistributions of source code must retain the above copyright notice, this list
+// of conditions and the following disclaimer.
+// Redistributions in binary form must reproduce the above copyright notice, this
+// list of conditions and the following disclaimer in the documentation and/or other
+// materials provided with the distribution.
+//
+// All advertising materials mentioning features or use of this software must display
+// the following acknowledgement: This product includes software developed by Polka Biome
+// Ltd., SORA, and Polkaswap.
+//
+// Neither the name of the Polka Biome Ltd. nor the names of its contributors may be used
+// to endorse or promote products derived from this software without specific prior written permission.
+
+// THIS SOFTWARE IS PROVIDED BY Polka Biome Ltd. AS IS AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL Polka Biome Ltd. BE LIABLE FOR ANY
+// DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+// BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use common::AssetInfoProvider;
+use common::AssetManager;
+use common::AssetRegulator;
+use common::AssetType;
+use common::Balance;
+use common::{AccountIdOf, AssetIdOf};
+use frame_support::pallet_prelude::*;
+use frame_system::pallet_prelude::*;
+use sp_runtime::traits::Zero;
+use sp_runtime::Saturating;
+use sp_std::collections::btree_set::BTreeSet;
+use sp_std::prelude::*;
+pub use weights::WeightInfo;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+mod benchmarking;
+
+pub mod weights;
+
+pub use pallet::*;
+
+#[derive(RuntimeDebug, Encode, Decode, TypeInfo, Clone, Copy, PartialEq, Eq)]
+pub enum Extension<BlockNumber, AssetId> {
+    MaxAmount(Balance),
+    RequiredProduct(AssetId, Balance),
+    DisallowedProduct(AssetId),
+    Expirable(BlockNumber),
+}
+
+#[derive(RuntimeDebug, Encode, Decode, TypeInfo, Clone, PartialEq, Eq)]
+pub struct Product<BlockNumber, AssetId> {
+    price_asset: AssetId,
```

Comment:
What kind of assets are allowed for exchange? All assets? Just regular assets? Can user define NFT or Soulbound asset for exchange?
Looks like we need to check this asset.